### ISSUE TITLE
False Positive on or_fun_call lint with enum constructor.

### ISSUE THREAD

**Issue by CryZe** [rust-lang/rust-clippy#1338]
`or_fn_call` shouldn't warn in case the function is just an enum constructor wrapping over local variables.

Here's the code:
```rust
pub enum Error<'a> {
    Path(&'a str),
}

pub fn question_mark(path: &str) -> Result<(), Error> {
    None.ok_or(Error::Path(path))?;
    
    Ok(())
}

pub fn try(path: &str) -> Result<(), Error> {
    try!(None.ok_or(Error::Path(path)));
    
    Ok(())
}
```

Warning:
```
Compiling playground v0.0.1 (file:///playground)
warning: use of `ok_or` followed by a function call, #[warn(or_fun_call)] on by default
 --> src/main.rs:6:5
  |
6 |     None.ok_or(Error::Path(path))?;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
help: try this
  |     None.ok_or_else(|| Error::Path(path))?;
  = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#or_fun_call
```

Here's the link to the code in the custom playground: [Playground](http://play.integer32.com/?gist=05b21e04af8e17cef2c158fd321722aa&version=nightly)

**Comment by Manishearth** [comment]
> cause it didn't do that before with the try! macro, so something is inconsistent here.

A lot of the clippy lints get turned off within macros.

**Comment by Manishearth** [comment]
This isn't a false positive fwiw, since clippy has no idea how expensive that function call is.

**Comment by CryZe** [comment]
Can it not check that it's just an enum constructor?

**Comment by Manishearth** [comment]
Fair point. I'm editing the bug report.

**Comment by llogiq** [comment]
So how could we determine something is a struct/enum constructor? fn name == return type name?

**Comment by oli-obk** [comment]
here's how we do it in miri: https://github.com/solson/miri/blob/ac2bf50f9d853a56ff1a236d456619b2d3b65b64/src/terminator/mod.rs#L236

**Comment by ishitatsuyuki** [comment]
This also applies to `Some`, which should instead pointed to `unwrap_or()`:

`None.or(Some(x))`

**Comment by Manishearth** [comment]
This should have been fixed by #919, but wasn't, because that only deals with const promotable things. 

I'll make a PR.
